src/builder/QuarkBuilder.sol

```diff
@@ -120,9 +128,28 @@ contract QuarkBuilder {
                         revert TooManyBridgeOperations();
                     }
 
-                    uint256 amountToBridge = srcAccountBalances[j].balance >= amountLeftToBridge
-                        ? amountLeftToBridge
-                        : srcAccountBalances[j].balance;
+                    uint256 amountToBridge;
+                    // Handle differently if the transfer intent token is the payment token
+                    if (payment.isToken && Strings.stringEqIgnoreCase(payment.currency, transferIntent.assetSymbol)) {
+                        // Apply payment token logics to figure out the amount to bridge
+                        if (
+                            srcAccountBalances[j].balance
+                                >= amountLeftToBridge + PaymentInfo.findMaxCost(payment, srcChainAccounts.chainId)
+                        ) {
+                            amountToBridge = amountLeftToBridge;
+                        } else {
+                            amountToBridge = srcAccountBalances[j].balance
+                                - PaymentInfo.findMaxCost(payment, srcChainAccounts.chainId);
+                        }
+                    } else {
+                        // Apply straighforward logics to bridge the token right away
+                        if (srcAccountBalances[j].balance >= amountLeftToBridge) {
+                            amountToBridge = amountLeftToBridge;
+                        } else {
+                            amountToBridge = srcAccountBalances[j].balance;
+                        }
+                    }
```

Comment:
Special handling + re-adjust for determining bridging amount when payment is payment. 
Because max cost is part of the balance of account's payment token.